core/frontend/src/Viewport.ts

```diff
@@ -2258,40 +2270,49 @@ export abstract class Viewport implements IDisposable {
         context.requestMissingTiles();
         target.changeScene(context.scene);
         isRedrawNeeded = true;
+        this._frameStatsCollector.endTime("createChangeSceneTime");
       }
 
       this._sceneValid = true;
     }
 
     if (!this._renderPlanValid) {
+      this._frameStatsCollector.beginTime("validateRenderPlanTime");
       this.validateRenderPlan();
+      this._frameStatsCollector.endTime("validateRenderPlanTime");
       isRedrawNeeded = true;
     }
 
     if (!this._decorationsValid) {
+      this._frameStatsCollector.beginTime("decorationsTime");
       const decorations = new Decorations();
       this.addDecorations(decorations);
       target.changeDecorations(decorations);
       this._decorationsValid = true;
       isRedrawNeeded = true;
+      this._frameStatsCollector.endTime("decorationsTime");
     }
 
+    this._frameStatsCollector.beginTime("flashTime");
     let requestNextAnimation = false;
     if (this.processFlash()) {
       target.setFlashed(undefined !== this._flashedElem ? this._flashedElem : Id64.invalid, this.flashIntensity);
       isRedrawNeeded = true;
       requestNextAnimation = undefined !== this._flashedElem;
     }
+    this._frameStatsCollector.endTime("flashTime");
 
     target.onBeforeRender(this, (redraw: boolean) => {

```

Comment:
Record this - this is one of the places Open Cities Planner hooks.

Reply:
Can you record IModelFrameLifecycle.onRenderOpaque in SceneCompositor.ts? That's where OCP does their heaviest lifting.

Reply:
Yes, it should be simple to wrap `IModelFrameLifecycle.onRenderOpaque.raiseEvent` with timing tracking in SceneCompositor.ts.

What line were you originally referring to?  `target.onBeforeRender()` in Viewport.ts?

Reply:
Comments appear directly below the line they reference - in this case line 2305 in Viewport.ts.

Reply:
Thanks for the feedback.  I have added 'onBeforeRenderTime' and 'onRenderOpaqueTime'.